[#65770] Document how to configure SCIM clients

ba1ash opened the pull request:
https://community.openproject.org/wp/65770
https://community.openproject.org/wp/65845
https://community.openproject.org/wp/65292
https://community.openproject.org/wp/62246

ba1ash committed:
[#65770] Document how to configure SCIM clients

https://community.openproject.org/work_packages/65770

Co-Authored-by: Jan Sandbrink <j.sandbrink@openproject.com>

NobodysNightmare commented on app/views/admin/scim_clients/edit.html.erb:
```diff
@@ -32,7 +32,7 @@ See COPYRIGHT and LICENSE files for more details.
   render(Primer::OpenProject::PageHeader.new) do |header|
     header.with_title { @scim_client.name }
-    header.with_description { t("admin.scim_clients.form.description_html", docs_url: "#") } # TODO: Deep link into documentation
+    header.with_description { t("admin.scim_clients.form.description_html", docs_url: ::OpenProject::Static::Links[:sysadmin_docs][:scim][:href]) }
```
It was more than one location that needed to be covered here.

One other is in `new.html.erb` and the last one in `app/forms/scim_clients/form.rb`.

MayaBerd replied:
@ba1ash this comment is still relevant

NobodysNightmare replied:
As far as I can tell `new.html.erb` is still missing.

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+description: SCIM
+keywords: SCIM
+---
+# SCIM (Enterprise add-on)
```
What do you think about replacing the title (including menu entries) with "SCIM API" instead of only "SCIM"?

To me only having a single "word" looks a bit too short. "SCIM API" at least indicates that this is about an API that OpenProject is offering.

ba1ash replied:
I would like **SCIM provisioning** orsomething like **SCIM user and group provisioning** more. WDYT?

NobodysNightmare replied:
What I like about that is that "provisioning" is a real word (same as "user and group"), where API is just the next abbreviation. Unless @MayaBerd has opposing thoughts, I think that "SCIM provisioning" is good for documentation headlines.

I hope that this is still consistent enough with the rest of our UI (menu entries in the application called **SCIM clients** etc).

MayaBerd replied:
I have no opposing thoughts, let's use SCIM provisioning. We'll adjust other website and release content to stay consistent.

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+> [!IMPORTANT] 
+> SCIM is an Enterprise add-on. If you do not see the button you will have to activate the Enterprise edition first.
+
+To activate and configure SCIM in OpenProject, navigate to *Administration* -> *Authentication* and choose -> *SCIM*.
```
This is not what the menu entry is called:

```suggestion
To activate and configure SCIM in OpenProject, navigate to *Administration* -> *Authentication* and choose -> *SCIM clients*.
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+# SCIM (Enterprise add-on)
+
+> [!IMPORTANT] 
+> SCIM is an Enterprise add-on. If you do not see the button you will have to activate the Enterprise edition first.
```
I guess "the button" refers to the "SCIM clients" menu entry. If you are not on enterprise corporate, you do see the button, but it has the enterprise icon next to it and clicking it will show you an enterprise banner.

CC @MayaBerd, I am pretty sure that we have standard wording for this documentation banner. The relevant enterprise edition here is Enterprise Corporate.

MayaBerd replied:
@NobodysNightmare yes, we do. I'll do my review round after Pavel had a chance to integrate your comments, to avoid conflicts.

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+![Add SCIM client. Index page creation button.](add_scim_1.png) 
+
+
+You are able to configure the following options for your SCIM client:
```
This intro doesn't match the list that comes afterwards. The intro indicates that configuration options will be listed, but the list then makes imperative statements ("Enter X", "Choose Y").

I think both kinds of guides can work, but the current list should rather be introduced with something like

```suggestion
Configure your SCIM client in the following form:
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+   ![Add SCIM client. Creation form. Name field.](add_scim_2.png) 
+2. Choose an **Authentication provider**. 
+   This is the service that users added by the SCIM provider will use to authenticate in OpenProject.
+   It should be available already in OpenProject. It can be [OIDC provider](../system-admin-guide/authentication/openid-providers/) or [SAML provider](../system-admin-guide/authentication/saml/).
```
"Should be" is ambiguous here. It could mean that if the user followed the steps outlined, then they _should_ now see such a service. But that's not what we mean here, because we have no clue whether they created one already.

I'd rather go for "must", to indicate that they'd better make sure this is true:

```suggestion
   It must have been configured before creating the SCIM client. It can be an [OIDC provider](../system-admin-guide/authentication/openid-providers/) or a [SAML provider](../system-admin-guide/authentication/saml/).
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+   This is how the SCIM client authenticates at OpenProject. Please ensure that OAuth tokens include the **scim_v2** scope.
+   There are three option:
+   a. **Static access token**
+      This is the most commomnly used authentication method for SCIM clients. In this case after clicking **Create** you get an access token that should be put to the SCIM client configuration on the other end.
```
```suggestion
      This is the most commonly used authentication method for SCIM clients. In this case after clicking **Create** you get an access token that should be put to the SCIM client configuration on the other end.
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+      There is generated access token. After closing the dialog with generated token you will not see it anymore.
+      ![Add SCIM client. Creation form. Static access token. Copy token.](add_scim_5.png) 
+      Token can be revoked. 
+      ![Add SCIM client. Creation form. Static access token. Revoke token.](add_scim_6.png) 
```
An important hint about static access tokens is that they are valid for one year and must be replaced manually.
IMO we should indicate that in the docs as well.

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,50 @@
+      There are generated client id and client secret. After closing the dialog with not see client secret anymore. 
+      ![Add SCIM client. Creation form. Client credentials. Copy client credentials.](add_scim_8.png) 
+   c. **JWT from identity provider**
+      In this case you have to specify **Subject claim** that authentication JWT contains.
```
There is no explanation what this authentication method even means.

MayaBerd replied:
@ba1ash would you like to add some information here?

MayaBerd committed:
update the new SCIM docs page

MayaBerd committed:
fix links

ba1ash committed:
[#65770] Update SCIM client JWT base auth docs.

ba1ash committed:
[#65770] Add a note to SCIM docs.

`user_acounts_deletable_by_admins` setting must be enabled for users deletion to work.

NobodysNightmare commented on app/forms/scim_clients/form.rb:
```diff
@@ -79,7 +79,7 @@ class Form < ApplicationForm
           label: ScimClient.human_attribute_name(:jwt_sub),
           required: true,
-          caption: I18n.t("admin.scim_clients.form.jwt_sub_description_html", docs_url: "#").html_safe, # TODO: correct docs url
+          caption: I18n.t("admin.scim_clients.form.jwt_sub_description_html", docs_url: ::OpenProject::Static::Links[:sysadmin_docs][:scim_jwt_authetication_method][:href]).html_safe,
```
:yellow_circle: Something that I just noticed now: There is a helper to access links that you can use here and for all other references to docs:

```suggestion
          caption: I18n.t("admin.scim_clients.form.jwt_sub_description_html", docs_url: ::OpenProject::Static::Links.url_for(:sysadmin_docs, :scim_jwt_authetication_method)).html_safe,
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,105 @@
+OpenProject supports automated user synchronization via SCIM API, enabling  seamless integration with your identity provider. Simplify and secure  user provisioning and de-provisioning while ensuring accurate user data  across systems.
+
+> [!NOTE] 
+> SCIM provisioning is an Enterprise add-on. [Click here for more information](https://www.openproject.org/enterprise-edition/) on the OpenProject Enterprise edition.
```
@MayaBerd Asking once more, since you approved the PR already: Do we not need to mention the specific enterprise plan in this banner? I'd have expected to read "Corporate" somewhere. This is the text that we have for single sign-on to Nextcloud storages:

> This advanced feature is an Enterprise add-on and can only be used with [Enterprise cloud](https://www.openproject-edge.com/docs/enterprise-guide/enterprise-cloud-guide) or [Enterprise on-premises](https://www.openproject-edge.com/docs/enterprise-guide/enterprise-on-premises-guide). It is available for all installations under the Enterprise Corporate plan. An upgrade from the free community edition is easy and helps support OpenProject.

MayaBerd replied:
@NobodysNightmare I deliberetely kept it generic for now. @as-op is working on an adjustment across all the docs, so that an enterprise plan will be included automatically for every enterprise feature.

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,105 @@
+
+Click on the client name to open the detailed view, edit the information, add revoke or add tokens. You will be able to edit the client information and tokens. 
+
+SCIM client tokens can be revoked. To revoke a token click the **Remove** icon at the far right end of the token listing. To add a new token click the **+ Token** button at the bottom of *Tokens* section.
```
I am pretty sure that the icon title is "Revoke":

```suggestion
SCIM client tokens can be revoked. To revoke a token click the **Revoke** icon at the far right end of the token listing. To add a new token click the **+ Token** button at the bottom of *Tokens* section.
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,105 @@
+
+#### b. **OAuth 2.0 client credentials**
+
+If in [Step 3](#step-3-choose-an-authentication-method) you selected **OAuth 2.0 client credentials**, after clicking **Create** you will get client credentials of newly created [OpenProject OAuth Application](../oauth-applications/#oauth-applications). These credentials should be entered into the SCIM client configuration on the other end. Then SCIM client is supposed to use provided client cred ...
```
at the beginning of step 3 we write

> Please ensure that OAuth tokens include the **scim_v2** scope.

But to be honest, here would be the right place to emphasize what we mean with that (maybe even removing the sentence from the start of step 3, if we do):

```suggestion
If in [Step 3](#step-3-choose-an-authentication-method) you selected **OAuth 2.0 client credentials**, after clicking **Create** you will get client credentials of newly created [OpenProject OAuth Application](../oauth-applications/#oauth-applications). These credentials should be entered into the SCIM client configuration on the other end. The SCIM client is supposed to use the provided client credentials to obtain an access token with the `scim_v2` scope from OpenProject and then use the access token in SCIM API requests.
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,105 @@
+This authentication method is intended exclusively for use with OpenID Connect setups.
+The SCIM client must be able to obtain a JWT from the OpenID Connect provider (e.g., Keycloak) that meets all of the following conditions:
+1. The `sub`  claim contains the value specified in the Subject Claim configuration.
+2. The `aud`  claim includes the Client ID from your OpenID Connect provider configuration.
```
```suggestion
2. The `aud`  claim includes OpenProject's Client ID from your OpenID Connect provider configuration.
```

NobodysNightmare commented on docs/system-admin-guide/authentication/scim/README.md:
```diff
@@ -0,0 +1,105 @@
+The SCIM client must be able to obtain a JWT from the OpenID Connect provider (e.g., Keycloak) that meets all of the following conditions:
+1. The `sub`  claim contains the value specified in the Subject Claim configuration.
+2. The `aud`  claim includes the Client ID from your OpenID Connect provider configuration.
+3. The `scope` claim includes `scim_v2`.
```
I think the three step listing above is good. I'd still add some flavour text to fill this requirements list with life. I.e. make it clear how this could be achieved.

Idea:

> This could for example be achieved by performing a client credentials token request towards the identity provider. In the case of Keycloak as an IDP, the `sub` claim would then equal to the UUID of the service account that's associated to the client obtaining the token.

ba1ash committed:
[#65770] Adjust SCIM provisioning docs.

- Use static links helper
- Add a Keycloak example for jwt based auth case

Co-Authored-by: Jan Sandbrink <j.sandbrink@openproject.com>
